### PR TITLE
ChangeFeedFixForDuplicateProcessing

### PR DESCRIPTION
**Pull request by xinlian12**
This is targeting to fix an issue where: multiple threads on the same processor are running parallel and processing the same partition.

### CODE REVIEW

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/implementation/WorkerTask.java [hunk 1]
```diff
@@ -2,34 +2,44 @@
...
-        try {
-            job.run();
-            logger.info("Partition controller worker task {} has finished running.", lease.getLeaseToken());
-        } finally {
-            logger.info("Partition controller worker task {} has exited.", lease.getLeaseToken());
-            job = null;
-            this.done = true;
-        }
+        job
+            .doOnSuccess(avoid-> logger.info("Partition controller worker task {} has finished running.", lease.getLeaseToken()))
```
Comment (mbhaskar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/implementation/WorkerTask.java:30]: nit: aVoid ->

Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/implementation/WorkerTask.java]: Thanks, fixed.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/CancellationToken.java [hunk 2]
```diff
@@ -2,20 +2,28 @@
...
 public class CancellationToken {
     private final CancellationTokenSource tokenSource;
+    private AtomicBoolean cancellationRequested;
 
     public CancellationToken(CancellationTokenSource source) {
         this.tokenSource = source;
+        this.cancellationRequested = new AtomicBoolean(false);
+    }
+
+    public void cancel() {
```
Comment (milismsft) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/CancellationToken.java:19]: This is not needed, please remove; the cancelation of a token should only be done through the CancellationTokenSource.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/CancellationToken.java [hunk 3]
```diff
@@ -2,20 +2,28 @@
...
+        this.cancellationRequested.set(true);
     }
 
     /**
      * @return true if the cancellation was requested from the source.
      */
     public boolean isCancellationRequested() {
-        return tokenSource.isCancellationRequested();
+        return tokenSource.isCancellationRequested() || this.cancellationRequested.get();
```
Comment (milismsft) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/CancellationToken.java:27]: Not needed

**Review by kushagraThapar — APPROVED** [review]
LGTM, thanks @xinlian12 for the amazing fix.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/implementation/PartitionControllerImpl.java [hunk 4]
```diff
@@ -109,56 +109,60 @@ public PartitionControllerImpl(
     }
 
     private Mono<Void> removeLease(Lease lease) {
-            if (this.currentlyOwnedPartitions.get(lease.getLeaseToken()) != null) {
-                WorkerTask workerTask = this.currentlyOwnedPartitions.remove(lease.getLeaseToken());
-
+        return Mono.justOrEmpty(this.currentlyOwnedPartitions.remove(lease.getLeaseToken()))
```
Comment (milismsft) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/implementation/PartitionControllerImpl.java:112]: use Mono.just(this) to handle the WorkTask removal from the concurrent list and pipe it to a flatMap that will do the lease update (see leaseManager.release()).

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/implementation/PartitionControllerImpl.java [hunk 5]
```diff
@@ -109,56 +109,60 @@ public PartitionControllerImpl(
...
+            .flatMap(workerTask -> {
```
Comment (milismsft) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/implementation/PartitionControllerImpl.java:113]: change the WorkerTask and add the following method:
`    public void cancelJob() {
        shutdownCts.cancel();
        this.interrupt();
    }
`

Also please change the "done" boolean to an AtomicBoolean or volatile:
`    private volatile boolean done = false;
`

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/CancellationToken.java (hunk 2 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/CancellationToken.java]: Chatted offline, will removed this and use cancellationTokenSource to cancel.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/CancellationToken.java (hunk 3 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/CancellationToken.java]: Removed.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/implementation/PartitionControllerImpl.java (hunk 4 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/implementation/PartitionControllerImpl.java]: Chatted offline, updated.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/implementation/PartitionControllerImpl.java (hunk 5 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/implementation/PartitionControllerImpl.java]: originally was doing override for interrupt method, changed to the suggested way: cancelJob. it is clearer.

**Review by moderakh — APPROVED** [review]
LGTM.

Please ask for @milismsft @FabianMeiswinkel sign off who are better familiar with this codepath.

**Review by milismsft — APPROVED** [review]

**Review by milismsft — APPROVED** [review]
looks great; thank you so much Annie 🥇

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/implementation/PartitionControllerImpl.java [hunk 6]
```diff
@@ -109,56 +109,61 @@ public PartitionControllerImpl(
...
     }
 
+    private Mono<Void> getWorkerJob(PartitionSupervisor partitionSupervisor, Lease lease, CancellationToken cancellationToken) {
+        return partitionSupervisor.run(cancellationToken)
+            .onErrorResume(throwable -> {
+                if (throwable instanceof PartitionSplitException) {
+                    PartitionSplitException ex = (PartitionSplitException) throwable;
+                    return this.handleSplit(lease, ex.getLastContinuation());
+                } else if (throwable instanceof TaskCancelledException) {
+                    logger.debug("Partition {}: processing canceled.", lease.getLeaseToken());
```
Comment (FabianMeiswinkel) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/implementation/PartitionControllerImpl.java:157]: info? Given how verbose we are in CFP on info level and how much it was needed to have the logs to understand what is going on?

**Review by FabianMeiswinkel — APPROVED** [review]
LGTM - thanks Annie and Millis

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/implementation/PartitionControllerImpl.java (hunk 6 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/changefeed/implementation/PartitionControllerImpl.java:157]: never mind - not needed here...

File: sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/rx/ChangeFeedProcessorTest.java [hunk 7]
```diff
@@ -64,7 +64,7 @@
 //    private final String hostName = "TestHost1";
     private final String hostName = RandomStringUtils.randomAlphabetic(6);
     private final int FEED_COUNT = 10;
-    private final int CHANGE_FEED_PROCESSOR_TIMEOUT = 5000;
+    private final int CHANGE_FEED_PROCESSOR_TIMEOUT = 5500;
```
Comment (milismsft) [inline sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/rx/ChangeFeedProcessorTest.java:67]: 7000? Half of a second increase might not be enough...

Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/src/test/java/com/azure/cosmos/rx/ChangeFeedProcessorTest.java]: increased to 7000

**Review by milismsft — APPROVED** [review]

**Review by milismsft — APPROVED** [review]

**Pull request merged**